### PR TITLE
feat: Support output.clean

### PR DESCRIPTION
**Pull request by suxin2017**
## Summary

[[Feature Request]: Support output.clean](https://github.com/web-infra-dev/rspack/issues/2577)

## Related issue (if exists)

Closes https://github.com/web-infra-dev/rspack/issues/2577

## Types of changes

- [ ] Docs change / Dependency upgrade
- [ ] Bug fix
- [x] New feature / Improvement
- [ ] Refactoring
- [ ] Breaking change

## Checklist

- [ ] I have added changeset via `pnpm run changeset`.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: crates/rspack_testing/src/test_config.rs [hunk 1]
```diff
@@ -218,6 +222,8 @@ impl From<PxToRem> for PxToRemOptions {
 #[derive(Debug, JsonSchema, Deserialize)]
 #[serde(rename_all = "camelCase", deny_unknown_fields)]
 pub struct Output {
+  #[serde(default = "false_by_default")]
```
Comment (hyf0) [inline crates/rspack_testing/src/test_config.rs:225]: If you want to use `false ` as the default value, you could just write `#[serde(default)]`.

File: crates/rspack_testing/src/test_config.rs [hunk 2]
```diff
@@ -62,6 +62,10 @@ fn true_by_default() -> bool {
   true
 }
 
+fn false_by_default() -> bool {
```
Comment (hyf0) [inline crates/rspack_testing/src/test_config.rs:65]: This seems redundant.

**Review by hyf0 — CHANGES_REQUESTED** [review]
@h-a-n-a. Could you take a look at the changes in `rspack_fs`?

File: crates/rspack_build/fixture/test.config.json [hunk 3]
```diff
@@ -1,7 +1,12 @@
...
   "entry": {
     "index": {
-      "import": ["./app.js"]
+      "import": [
+        "./app.js"
+      ]
     }
+  },
+  "output": {
+    "clear": true
```
Comment (ahabhgk) [inline crates/rspack_build/fixture/test.config.json:10]: tests should be added in `packages/rspack/tests`

File: crates/rspack_binding_options/src/options/raw_output.rs [hunk 4]
```diff
@@ -132,6 +133,7 @@ impl RawOptionsApply for RawOutputOptions {
 
     Ok(OutputOptions {
       path: self.path.into(),
+      clear: self.clear.into(),
```
Comment (IWANABETHATGUY) [inline crates/rspack_binding_options/src/options/raw_output.rs:136]: `self.clear` is enough.

File: crates/rspack_core/src/compiler/mod.rs [hunk 5]
```diff
@@ -182,15 +182,18 @@ where
 
   #[instrument(name = "emit_assets", skip_all)]
   pub async fn emit_assets(&mut self) -> Result<()> {
+    let output_path = &self.options.output.path;
+
+    if self.options.output.clear {
+      self.output_filesystem.rm(output_path).await?;
+    }
```
Comment (IWANABETHATGUY) [inline crates/rspack_core/src/compiler/mod.rs:189]:
```suggestion
    if self.options.output.clear {
      self.output_filesystem.rm(&self.options.output.path).await?;
    }
```

Comment (IWANABETHATGUY) [reply crates/rspack_core/src/compiler/mod.rs:189]: NIT

File: crates/rspack_fs/src/async.rs [hunk 6]
```diff
@@ -23,6 +23,8 @@ pub trait AsyncWritableFileSystem {
   /// Write a slice as the entire contents of a file.
   /// This function will create a file if it does not exist, and will entirely replace its contents if it does.
   fn write<P: AsRef<Path>, D: AsRef<[u8]>>(&self, file: P, data: D) -> BoxFuture<'_, Result<()>>;
+
+  fn rm<P: AsRef<Path>>(&self, dir: P) -> BoxFuture<'_, Result<()>>;
```
Comment (h-a-n-a) [inline crates/rspack_fs/src/async.rs:27]: I would in favor of the name provided by the original std impl. Could you change it to `remove_dir_all`?

Comment (suxin2017) [reply crates/rspack_fs/src/async.rs]: Thank you, it has been modified

File: crates/rspack_fs/src/native.rs [hunk 7]
```diff
@@ -57,6 +57,12 @@ cfg_async! {
       let fut = async move { tokio::fs::write(file, data).await.map_err(Error::from) };
       Box::pin(fut)
     }
+
+    fn remove_dir_all<P: AsRef<Path>>(&self, dir: P) -> BoxFuture<'_, Result<()>> {
+      let dir = dir.as_ref().to_string_lossy().to_string();
+      let fut = async move { tokio::fs::remove_dir_all(dir).await.map_err(Error::from) };
```
Comment (hyf0) [inline crates/rspack_fs/src/native.rs:63]: I believe [tokio::fs::remove_dir_all](https://docs.rs/tokio/0.3.1/tokio/fs/fn.remove_dir_all.html) is ok to accept `Path` as arguments. You should be able to just pass `dir` `to tokio::fs::remove_dir_all`.

**Review by hyf0 — CHANGES_REQUESTED** [review]

File: crates/rspack_build/fixture/test.config.json (hunk 3 above)
Comment (suxin2017) [reply crates/rspack_build/fixture/test.config.json]: I put the files in the packages/rspack/tests/Output.test.ts

File: crates/node_binding/binding.d.ts [hunk 8]
```diff
@@ -12,6 +12,7 @@ export interface ThreadsafeNodeFS {
   writeFile: (...args: any[]) => any
   mkdir: (...args: any[]) => any
   mkdirp: (...args: any[]) => any
+  rm: (...args: any[]) => any
```
Comment (hyf0) [inline crates/node_binding/binding.d.ts:15]: You need to update this file by running `pnpm -w run build:cli:debug`

File: crates/rspack_fs_node/src/node.rs [hunk 9]
```diff
@@ -67,6 +67,7 @@ cfg_async! {
     pub write_file: JsFunction,
     pub mkdir: JsFunction,
     pub mkdirp: JsFunction,
+    pub rm: JsFunction,
```
Comment (hyf0) [inline crates/rspack_fs_node/src/node.rs:70]: could you also change the `rm` here to `remove_dir_all`?

File: crates/rspack_fs_node/src/node.rs [hunk 10]
```diff
@@ -157,5 +159,6 @@ cfg_async! {
     pub(crate) write_file: ThreadsafeFunction<(String, Vec<u8>), ()>,
     pub(crate) mkdir: ThreadsafeFunction<String, ()>,
     pub(crate) mkdirp: ThreadsafeFunction<String, Either<String, ()>>,
+    pub(crate) rm: ThreadsafeFunction<String, Either<String, ()>>,
```
Comment (hyf0) [inline crates/rspack_fs_node/src/node.rs:162]: `remove_dir_all`

File: packages/rspack/src/fileSystem.ts [hunk 11]
```diff
@@ -1,7 +1,10 @@
+import { join } from "path";
+
 export interface ThreadsafeWritableNodeFS {
 	writeFile: (...args: any[]) => any;
 	mkdir: (...args: any[]) => any;
 	mkdirp: (...args: any[]) => any;
+	rm: (...args: any[]) => any;
```
Comment (hyf0) [inline packages/rspack/src/fileSystem.ts:7]: `remove_dir_all`

File: packages/rspack/src/fileSystem.ts [hunk 12]
```diff
@@ -13,8 +16,30 @@ function createThreadsafeNodeFSFromRaw(
 		mkdirp: dir =>
 			fs.mkdirSync(dir, {
 				recursive: true
-			})
+			}),
+		rm: dir => {
```
Comment (hyf0) [inline packages/rspack/src/fileSystem.ts:20]: `remove_dir_all`

**Review by hyf0 — CHANGES_REQUESTED** [review]

File: crates/node_binding/binding.d.ts (hunk 8 above)
Comment (suxin2017) [reply crates/node_binding/binding.d.ts]: done

File: crates/rspack_fs_node/src/node.rs (hunk 9 above)
Comment (suxin2017) [reply crates/rspack_fs_node/src/node.rs]: done

**Review by hyf0 — APPROVED** [review]
LGTM. You might need resolve the conflicts.

**Comment by suxin2017** [comment]
> LGTM. You might need resolve the conflicts. @h-a-n-a cc

done

**Review by hyf0 — APPROVED** [review]

**Review by hyf0 — APPROVED** [review]

**Review by hardfist — APPROVED** [review]

**Comment by GarrisonD** [comment]
It's missing in the docs 😢 I was lucky to read changelogs and notice that `rspack` supports it now out of the box. Otherwise I would still do `rm -rf dist && rspack build` 😢

**Comment by hyf0** [comment]
> It's missing in the docs 😢 I was lucky to read changelogs and notice that `rspack` supports it now out of the box. Otherwise I would still do `rm -rf dist && rspack build` 😢

Sorry for that. We have created https://github.com/web-infra-dev/rspack-website/issues/265 but haven't resolved it.

**Pull request merged**